### PR TITLE
Add Table Transformer

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds [Table Transformer](https://github.com/microsoft/table-transformer) by Microsoft, which are DETR-compatible models for table detection and table structure recognition tasks in unstructured documents.

Note: I'm making some updates to the original DETR implementation, however these are justified by the fact that the original DETR implementation by Facebook AI also includes these things, which I didn't add when first porting DETR. Hence, our DETR implementation is now more aligned with the original one.

To do:

- [ ] transfer checkpoints to the Microsoft organization
- [ ] add link to notebook

### CODE REVIEW

**Comment by sgugger** [comment]
I'm very much not in favor of adding a new config parameter that controls where the layernorm is applied. I'm not surprised the original code has it, as Facebook AI usually codes models in a modular way, but not Transformers. We had the same thing with BART and friends, and they are coded as distinct models in the library.

File: docs/source/en/model_doc/table_transformer.mdx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+specific language governing permissions and limitations under the License.
+-->
+
+# Table Transformer
+
+## Overview
+
+The Table Transformer model was proposed in [PubTables-1M: Towards comprehensive table extraction from unstructured documents](https://arxiv.org/abs/2110.00061) by
+Brandon Smock, Rohith Pesala, Robin Abraham. The authors introduce a new dataset, PubTables-1M, to benchmark progress in table extraction from unstructured documents,
+as well as table structure recognition and functional analysis. The authors train 2 [DETR](detr) models, one for table detection and one for table structure recognition, dubbed Table Transformers.
```
Comment (alaradirik) [inline docs/source/en/model_doc/table_transformer.mdx:19]:
```suggestion
as well as table structure recognition and functional analysis tasks. The authors train two [DETR](detr) models, one for table detection and one for table structure recognition, dubbed Table Transformers.
```

File: docs/source/en/model_doc/table_transformer.mdx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+comprehensive dataset for table extraction, called PubTables-1M. PubTables-1M contains nearly one million tables from scientific articles, supports multiple input
+modalities, and contains detailed header and location information for table structures, making it useful for a wide variety of modeling approaches. It also addresses a significant
+source of ground truth inconsistency observed in prior datasets called oversegmentation, using a novel canonicalization procedure. We demonstrate that these improvements lead to a
+significant increase in training performance and a more reliable estimate of model performance at evaluation for table structure recognition. Further, we show that transformer-based
+object detection models trained on PubTables-1M produce excellent results for all three tasks of detection, structure recognition, and functional analysis without the need for any
+special customization for these tasks.*
+
+Tips:
+
+- The authors released 2 models, one for table detection in documents, one for table structure recognition (the task of recognizing the individual rows, columns etc. in a table).
```
Comment (alaradirik) [inline docs/source/en/model_doc/table_transformer.mdx:34]:
```suggestion
- The authors released two models, one for table detection in documents, one for table structure recognition (the task of recognizing the individual rows, columns etc. in a table).
```

File: src/transformers/models/table_transformer/convert_table_transformer_to_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    image = F.normalize(image, mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
+    return image
+
+
+@torch.no_grad()
+def convert_table_transformer_checkpoint(checkpoint_url, pytorch_dump_folder_path, push_to_hub):
+    """
+    Copy/paste/tweak model's weights to our DETR structure.
+    """
+
```
Comment (alaradirik) [inline src/transformers/models/table_transformer/convert_table_transformer_to_pytorch.py:194]:
```suggestion
```

**Review by alaradirik — COMMENTED** [review]
Looks good to me! My only concern is that this model might be difficult to publicize as the paper's main contribution is the PubTables-1M dataset and demonstrating how DETR can be used to solve table extraction and related tasks. 

Could we add the notebook you are working on to the notebooks repo?

**Comment by NielsRogge** [comment]
Closing this PR in favor of #19614

**Pull request closed without merging**